mitodl/open-discussions
# Search facet improvements

#### Pre-Flight checklist


#### What are the relevant tickets?
Partially addresses #1887

#### What's this PR do?
Implements the following front-end changes to the course search facets:
- [x] Display the hit count next to each facet
- [x] Don't display a facet for which there are zero hits in the current result set
- [x] Sort the facets by hit count
- [x] Only display the first 5 facet filters, with 'View more/less' functionality if there's > 5

Changes the search API call to include aggregations with each response, assuming facets are included with the request. 

Removes the separate aggregations request which used to be called on initial page load to get a static list of all facet choices.

#### How should this be manually tested?
- if you've already imported courses before, run `recreate_index` (there was a recent addition of 'availability' to the course docs).  If you haven't, run `./manage.py backpopulate_edx_data` and `./manage.py backpopulate_ocw_data` (interrupt the latter after about 10 minutes or it will continue for hours).
- Go to `/courses/search`
- Facets for subject area, platform, and availability should appear.  Only the first 5 subject areas should be visible along with a 'View more' link.
- Clicking 'View more' should show all the subjects and a 'View less' link, which if clicked just shows 5 again.
- Check one of the subject facets.  The search results and URL should change accordingly,  along with the facets and their counts.  The facets in each category should be sorted by count in descending order.
- Check one of the availability facets.  The other availability facets should disappear because they are not found in the results.
- Check one of the platform facets.  The other platform facet should disappear because it is not found in the results.
- Go to the regular search page and verify you can still search for posts, comments, and profiles.

#### Screenshots
<img width="1002" alt="screen shot 2019-03-07 at 4 26 03 pm" src="https://user-images.githubusercontent.com/187676/53990242-d66acf00-40f5-11e9-8e1e-b27616e62a08.png">

<img width="371" alt="screen shot 2019-03-07 at 4 26 16 pm" src="https://user-images.githubusercontent.com/187676/53990247-d965bf80-40f5-11e9-8086-b2fc4d606698.png">
